src/TuxxState.js

```diff
@@ -0,0 +1,340 @@
+        if (valAtKey !== null && typeof valAtKey === 'object' && !Array.isArray(valAtKey) && stAtKey !== null && typeof stAtKey === 'object' && !Array.isArray(stAtKey)) {
+          // shallow copy the keys from the currentState into our newState so that there are no references to the currentState keys, but our newState contains identical values
+          newSt[key] = assign({}, stAtKey);
+          // invoke recurseKeys with the nested objects inside the current key
+          recurseKeys(stAtKey, newSt[key], valAtKey);
+        } else {
+          // we have found a deep key in newProps and all keys match so far, so we invoke our callback, which sets the result of invoking our callback with these deep keys to be the value at the corresponding key in our newState
+          callback(newSt, key, stAtKey, valAtKey);
+        }
+      }
+    }
+  };
+  // invoke recurseKeys and build newState object by passing in the currentState, our empty newState, and newProps
+  recurseKeys(currentState, newState, newProps);
+  return newState;
+};
+
+// invariantNumberCheck FUNCTION: internal function that ensures a given value is a number or throws an error
+// @param input UNKNOWN: value we are performing invariant check upon
+var invariantNumberCheck = function (input) {
+  if (typeof(input) !== "number") {
+    invariant(!input, 'Cannot perform operation on "%s" because it is not of type number.', input);
+  };
+};
+
+// invariantArrayCheck FUNCTION: internal function that ensures a given value is an array or throws an error
+// @param input UNKNOWN: value we are performing invariant check upon
+var invariantArrayCheck = function (input) {
+  if (!Array.isArray(input)) {
+    invariant(!input, 'Cannot perform operation on "%s" because it is not an array.', input);
+  };
+};
+
+// invariantValueCheck FUNCTION: internal function that ensures a given value is not stored in an array or an object
+// @param input UNKNOWN: value we are performing invariant check upon
+var invariantValueCheck = function (input) {
+  if (typeof(input) === "object") {
+    invariant(!input, 'Cannot perform operation on "%s" because it must not be an array or object.', input);
+  }
+};
+
+// invariantNumberOrStringCheck FUNCTION: internal function that ensures a given value is a number or string or throws an error
+// @param input UNKNOWN: value we are performing invariant check upon
+var invariantNumberOrStringCheck = function (input) {
+  if (typeof(input) !== "number" && typeof(input) !== "string") {
+    invariant(!input, 'Cannot perform operation on "%s" because it is not of type number or of type string.', input);
+  };
+};
+
+// invariantArgCheck FUNCTION: internal function that ensures a given value is defined and is an object or throws an error
+// @param input UNKNOWN: value we are performing invariant check upon
+var invariantArgCheck = function (input) {
+  if (!input) {
+    invariant(input, 'This function requires an object as an argument.');
```

Comment:
Both of the invariant error messages are the same. Is there any way to combine these into one call? Won't the else if conditional check work for the if case as well?